homeassistant/components/esphome/bluetooth/scanner.py

```diff
@@ -61,50 +63,62 @@ def _async_expire_devices(self, _datetime: datetime.datetime) -> None:
             if now - timestamp > FALLBACK_MAXIMUM_STALE_ADVERTISEMENT_SECONDS
         ]
         for address in expired:
-            del self._discovered_devices[address]
+            del self._discovered_device_advertisement_datas[address]
             del self._discovered_device_timestamps[address]
 
     @property
     def discovered_devices(self) -> list[BLEDevice]:
         """Return a list of discovered devices."""
-        return list(self._discovered_devices.values())
+        return [
+            device_advertisement_data[0]
+            for device_advertisement_data in self._discovered_device_advertisement_datas.values()
+        ]
 
-    async def async_get_device_by_address(self, address: str) -> BLEDevice | None:
-        """Get a device by address."""
-        return self._discovered_devices.get(address)
+    @property
+    def discovered_devices_and_advertisement_data(
+        self,
+    ) -> dict[str, tuple[BLEDevice, AdvertisementData]]:
+        """Return a list of discovered devices and advertisement data."""
+        return self._discovered_device_advertisement_datas
 
     @callback
     def async_on_advertisement(self, adv: BluetoothLEAdvertisement) -> None:
         """Call the registered callback."""
         now = time.monotonic()
         address = ":".join(TWO_CHAR.findall("%012X" % adv.address))  # must be upper
         name = adv.name
-        if prev_discovery := self._discovered_devices.get(address):
+        if prev_discovery := self._discovered_device_advertisement_datas.get(address):
             # If the last discovery had the full local name
             # and this one doesn't, keep the old one as we
             # always want the full local name over the short one
-            if len(prev_discovery.name) > len(adv.name):
-                name = prev_discovery.name
+            prev_device = prev_discovery[0]
+            if len(prev_device.name) > len(adv.name):
+                name = prev_device.name
 
-        advertisement_data = AdvertisementData(  # type: ignore[no-untyped-call]
+        advertisement_data = AdvertisementData(
             local_name=None if name == "" else name,
             manufacturer_data=adv.manufacturer_data,
             service_data=adv.service_data,
             service_uuids=adv.service_uuids,
+            rssi=adv.rssi,
+            tx_power=-127,
+            platform_data=(),
         )
         device = BLEDevice(  # type: ignore[no-untyped-call]
             address=address,
             name=name,
             details=self._details,
-            rssi=adv.rssi,
```

Comment:
Wasnt this supposed to remain in device untill 0.20?

Reply:
It's in both for now and will start throwing warnings in 0.20.0

Reply:
Yes that is why it should still be passed here. Otherwise integrations relying on it breaks? Or is it added some other way to the device object?

Reply:
It's on the Bluetooth service info object so it's still accessible.  

I would prefer not to pass it though for now and only add it if we find something that breaks before release since we know it's going to break in bleak soon.

If we leave it in we may miss any unexpected usage of it and may find out too late when we don't have a window to add it back.

Reply:
I will just end up zero, so maybe it doesnt matter too much. I just feel like we should follow the soft deprecation as done in bleak on our own bleak scanner for esphome, but its no major deal. We can leave it as is.

Reply:
Since it is left in for the platform scanner, I left it in for ESPHome as well.  I don't think anything is using this other than for logging and sensors so any future breakage is likely low impact. It is really up to the library author using bleak to follow upstream changes anyways so if it breaks its on them.